### PR TITLE
[WIP] Add Wayland backend

### LINKED ISSUE #1243: Our own backends for X11/Wayland
There are way too many limitations and tradeoffs that GTK/GDK have to support both of them, while not allowing us to do things that we want to do. Notable examples are separate-thread rendering (we have to use our own X11 connection to do that) and issues with absolute window positioning on Wayland (Wayland has `xdg_surface_set_window_geometry` which isn't used by GTK for some reason which leads to our popups being broken).

There are some major caveats on that way:
1) System dialogs (open/save file, folder selection, etc). We need to either still load GTK or have them out-of-proc somehow. The out-of-proc way will cause issues with making dialogs modal to the parent window on X11
2) Input methods. Right now we are relying on GTK do support them
3) Client-side window decorations. Some of the wayland shells are requiring applications to draw window frame, title and buttons, we need to support that somehow

### LINKED ISSUE #2338: Chrome OS?
Being an early-days WPF fan, I'm really excited about this project!

Maybe a silly question, but I was wondering, is Chrome OS a supported target platform?

### PR DESCRIPTION
**Pull request by affederaffe**
## What does the pull request do?
Add a Wayland backend for Linux.
This PR is based on [the previous one](https://github.com/AvaloniaUI/Avalonia/pull/8352) with significant bug fixes and easier maintenance . 

In order to test this PR, locally building NWayland NuGet packages is required:
`git clone --recursive https://github.com/affederaffe/NWayland` next to the `Avalonia` directory

├── Avalonia
│   ├── samples
│   ├── src
│   ├── ...
├── NWayland
│   ├── src
│   ├── ...

Navigate to `./NWayland/src/NWayland.Scanner` and `dotnet run` to generate the bindings.
Then `dotnet build -c Release` the `./NWayland/src/NWayland` and `./NWayland/src/NWayland.Protocols.Plasma` projects, which creates the required NuGet packages.

These steps won't be necessary once NWayland is released officially on NuGet.

TODOs:
- Better CSD integration
- IME

## What is the current behavior?
XWayland is used in order to run Avalonia apps on Wayland systems.

## What is the updated/expected behavior with this PR?
Use Wayland instead of X11 on compatible systems.

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Breaking changes
None

## Obsoletions / Deprecations
None

## Fixed issues
Fixes #1243
Fixes #2338

### CODE REVIEW

**Comment by iMonZ** [comment]
@affederaffe Do you think this can be still merged for the 11 update?

**Comment by affederaffe** [comment]
There are still some minor bugs to be fixed before merging and since Avalonia 11 seems to be moving closer and closer to RCs (feature freeze), I don't think we'll see this merged before 11.0. 11.x seems to be possible though, see [the prior discussion](https://github.com/AvaloniaUI/Avalonia/pull/8352#issuecomment-1461342416)

**Comment by danwalmsley** [comment]
> There are still some minor bugs to be fixed before merging and since Avalonia 11 seems to be moving closer and closer to RCs (feature freeze), I don't think we'll see this merged before 11.0. 11.x seems to be possible though, see [the prior discussion](https://github.com/AvaloniaUI/Avalonia/pull/8352#issuecomment-1461342416)

11.1.

11.0 rc is coming in a matter of days and the team wont have time to test or review this in that time.

11.1 is likely to be very very soon after 11.0 however.

**Comment by jmacato** [comment]
@affederaffe seems like there's an error building NWayland.Protocols.Plasma

CSC : error CS7027: Error signing output with public key from file '/NWayland/src/sign.snk' -- Could not find file '/NWayland/src/sign.snk'. [/NWayland/src/NWayland.Protocols.Plasma/NWayland.Protocols.Plasma.csproj::TargetFramework=netstandard2.0]
CSC : error CS8102: Public signing was specified and requires a public key, but no public key was specified. [/NWayland/src/NWayland.Protocols.Plasma/NWayland.Protocols.Plasma.csproj::TargetFramework=netstandard2.0]

**Comment by affederaffe** [comment]
Please `git pull` in NWayland and try again, see https://github.com/affederaffe/NWayland/commit/c032e578c282a1d53dfed510c3fa3149a456b0be.
Seems like I forgot to push those changes

**Comment by jmacato** [comment]
seems like NWayland needs to be published first so that the unit tests would work again

File: Avalonia.sln [hunk 1]
```diff
@@ -274,6 +274,7 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Avalonia.Headless.XUnit.Uni
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "MobileSandbox.Browser", "samples\MobileSandbox.Browser\MobileSandbox.Browser.csproj", "{43FCC14E-EEBE-44B3-BCBC-F1C537EECBF8}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Avalonia.Wayland", "src\Avalonia.Wayland\Avalonia.Wayland.csproj", "{684E5DA6-07B9-4046-9984-0BA6F56FD39E}"
```
Comment (ptasev) [inline Avalonia.sln:277]: Recent merge swallowed an `EndProject`

**Comment by kekekeks** [comment]
There are at least 2 blockers from my side:
- general support for CSD in the base framework (stealing pixels from ClientSize isn't a good idea)
- merge changes to NWayland and make it to be properly packaged

**Comment by jgcodes2020** [comment]
> There are at least 2 blockers from my side:
> 
>     * general support for CSD in the base framework (stealing pixels from ClientSize isn't a good idea)
> 
>     * merge changes to NWayland and make it to be properly packaged

I think CSD deserves its own tracker issue.

**Comment by jgcodes2020** [comment]
I took a closer look at the code, and I've noticed that this is still missing a `NativeControlHost` and access to platform handles.

This raises some issues:

- Wayland uses some global objects (`wl_compositor`, `wl_subcompositor`, `wl_display`). Should we expose Avalonia's instances directly?
- A `wl_surface` is generally not enough to get anything done&mdash;you also need the surface role (`xdg_toplevel`, `wl_subsurface`) to do certain things. How do we expose both `wl_surface` and `<surface role>`?

**Comment by maxkatz6** [comment]
@jgcodes2020 does wayland even support windows/control embedding?

**Comment by kekekeks** [comment]
@jgcodes2020
>I've noticed that this is still missing a NativeControlHost

1) Wayland doesn't support cross-client control embedding
2) It's not really possible to share a single connection between two toolkits (in some event handlers it's not clear which of the subscribers has ownership over the object created by libwayland-client and it's not possible to tell if there are subscriptions from another toolkit

The recommended  (from the official `#wayland` IRC channel) way to embed widgets from other toolkits is to "write your own nested wayland compositor" and ask said toolkit to use it instead of the system one.
Then you need to create a toplevel window in such toolkit and then find that window in your nested compositor (or maybe spawn a separate compositor per embedded control).
Then you are required to pass input events and receive rendered frames via completely compatible and similar DMA-BUF or EGLStreams APIs depending on the current graphics hardware. 
There is also no defined way to integrate TAB-navigation for input focus since the embedded control would think that it's a normal toplevel window, so you'll probably need to patch said third-party toolkit to provide one too.

Such nested compositor would be so rigorously and intuitively obvious to even the most casual software developer, that we leave the implementation to ones who want to use the wonderful, secure, and powerful Wayland protocol for any non-helloworld applications, they'll surely be able to do it an hour or two.

**Comment by jgcodes2020** [comment]
> @jgcodes2020 does wayland even support windows/control embedding?

Qt uses [xdg_foreign](https://wayland.app/protocols/xdg-foreign-unstable-v2) to do this.

**Comment by kekekeks** [comment]
@jgcodes2020
What a peculiar link you've posted here. It indeed describes a protocol that allows apps to exchange surface handles. However an attentive reader (in case said reader has read the protocol at all) would notice that it only works with xdg_toplevel surfaces to establish dialog parent relationships and has nothing to do with control embedding. 
It was introduced to provide system file dialogs for containerized apps by the means of xdg-portal.

**Comment by kekekeks** [comment]
@jgcodes2020 Just to clarify: this sarcastic tone isn't pointed at you but at wayland poor design decisions. It's really hard to do anything useful when protocol authors are actively trying to prevent you from that using made up reasons as excuses.

**Comment by kekekeks** [comment]
> some frameworks (e.g. SDL, GTK, VLC) support rendering to existing wl_surfaces

I don't think there is a way to make GTK to use a pre-existing wl_display connection. It also would be really hard to integrate event loops, given that Avalonia uses WPF's dispatcher model with rather strict guarantees about callback  execution ordering.

We'll need to think about SDL/VLC integration, since those probably expect being non-primary toolkits in the app.

**Comment by jgcodes2020** [comment]
> It's not really possible to share a single connection between two toolkits (in some event handlers it's not clear which of the subscribers has ownership over the object created by libwayland-client and it's not possible to tell if there are subscriptions from another toolkit

When it comes to "ownership" of a Wayland object there is [wl_proxy_set_tag](https://manpages.debian.org/experimental/libwayland-doc/wl_proxy.3.en.html#void_wl_proxy_set_tag_(struct_wl_proxy_*_proxy,_const_char_*const_*_tag)) but I don't know exactly how useful it is.

**Comment by jgcodes2020** [comment]
Adding on to some of my previous comments here: providing access to all Wayland singletons used by Avalonia (`wl_display`, `wl_compositor`, etc.) plus `wl_surface` handles is necessary for embedding, as the `wl_subsurface` has to be created on the same connection as the parent.

**Comment by iMonZ** [comment]
> There are at least 2 blockers from my side:
> 
> * general support for CSD in the base framework (stealing pixels from ClientSize isn't a good idea)
> * merge changes to NWayland and make it to be properly packaged

Still the same issues left?

**Comment by ErikApption** [comment]
> We'll need to think about SDL/VLC integration, since those probably expect being non-primary toolkits in the app.

Going to assume that this is wayland related - I'm using [the unofficial avalonia vlc with overlay](https://github.com/radiolondra/libvlcsharp/tree/3.x/src/LibVLCSharp.Avalonia.Unofficial) on Avalonia. Overlay/compositing/alpha works fine but when I exit the app the whole screen freezes and spits
```
gnome-shell[1685]: Window manager warning: Buggy client sent a _NET_ACTIVE_WINDOW message with a timestamp of 0 for 0xa0016f
mpp[35290]: mpp_info: mpp version: unknown mpp version for missing VCS info
gnome-shell[1685]: Window manager warning: Invalid WM_TRANSIENT_FOR window 0xa00169 specified for 0xa0016f.
```

Looking at the Avalonia Window source code and Ryujinx, I'm wondering if Avalonia needs to expose better Window close methods for native components.

File: Avalonia.sln (hunk 1 above)
Comment (ptasev) [reply Avalonia.sln:288]: This looks to be the case for Avalonia.Metal now

**Comment by affederaffe** [comment]
As I see it window embedding isn't really a problem anymore, at least on the video-player side.
phonon-vlc seems to be able to integrate vlc and mpv can also be used with an GL context iirc.

**Comment by IsaacMarovitz** [comment]
Any update on this PR?

**Comment by affederaffe** [comment]
There's still some work to do:
1. Client Side Decorations: Probably needs support in Avalonia.Base Window itself, as pointed out, stealing client side pixels isn't a good idea
2. Touch support: currently just crashes when receiving a touch
3. IME support: currently in progress
4. NWayland NuGet package (submopdules are tolerable for my own development, but this approach won't work for master)
5. For a smooth transition, 3rd-party-support, most important video players: needs to use the library directly without window embedding, see libMpv (for example, [mpvqt](https://github.com/KDE/mpvqt) and libVlc. Embedding support is just not feasible as stated above
6. (optional) survive compositor crashes
There's probably some more, those are just the things from the top of my head.

**Comment by cla-avalonia** [comment]
@affederaffe,

Please read the following Contributor License Agreement (CLA). If you agree with the CLA, please reply with the following:
```
@cla-avalonia agree
```

<details>
    <summary>Contributor License Agreement</summary>
    
### Contribution License Agreement

This Contribution License Agreement ( **“Agreement”** ) is agreed to by the party signing below ( **“You”** ),
and conveys certain license rights to the AvaloniaUI OÜ ( **“AvaloniaUI OÜ”** ) for Your contributions to
AvaloniaUI OÜ open source projects. This Agreement is effective as of the latest signature date below.

**1. Definitions.**

**“Code”** means the computer software code, whether in human-readable or machine-executable form,
that is delivered by You to AvaloniaUI OÜ under this Agreement.

**“Project”** means any of the projects owned or managed by AvaloniaUI OÜ and offered under a license
approved by the Open Source Initiative (www.opensource.org).

**“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
Project, including but not limited to communication on electronic mailing lists, source code control
systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
discussing and improving that Project, but excluding communication that is conspicuously marked or
otherwise designated in writing by You as “Not a Submission.”

**“Submission”** means the Code and any other copyrightable material Submitted by You, including any
associated comments and documentation.

**2. Your Submission.** You must agree to the terms of this Agreement before making a Submission to any
Project. This Agreement covers any and all Submissions that You, now or in the future (except as
described in Section 4 below), Submit to any Project.

**3. Originality of Work.**  You represent that each of Your Submissions is entirely Your 
original work. Should You wish to Submit materials that are not Your original work, 
You may Submit them separately to the Project if You (a) retain all copyright and 
license information that was in the materials as you received them, (b) in the 
description accompanying your Submission, include the phrase &quot;Submission 
containing materials of a third party:&quot; followed by the names of the third party and any 
licenses or other restrictions of which You are aware, and (c) follow any other 
instructions in the Project&#39;s written guidelines concerning Submissions.

**4. Your Employer.** References to “employer” in this Agreement include Your employer or anyone else
for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
Submission is made in the course of Your work for an employer or Your employer has intellectual
property rights in Your Submission by contract or applicable law, You must secure permission from Your
employer to make the Submission before signing this Agreement. In that case, the term “You” in this
Agreement will refer to You and the employer collectively. If You change employers in the future and
desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
and secure permission from the new employer before Submitting those Submissions.

**5. Licenses.**

**a. Copyright License.** You grant AvaloniaUI OÜ, and those who receive the Submission directly
or indirectly from AvaloniaUI OÜ, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable
license in the Submission to reproduce, prepare derivative works of, publicly display, publicly perform,
and distribute the Submission and such derivative works, and to sublicense any or all of the foregoing
rights to third parties.

**b. Patent License.** You grant AvaloniaUI OÜ, and those who receive the Submission directly or
indirectly from AvaloniaUI OÜ, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license
under Your patent claims that are necessarily infringed by the Submission or the combination of the
Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
import or otherwise dispose of the Submission alone or with the Project.

**c. Other Rights Reserved.** Each party reserves all rights not expressly granted in this Agreement.
No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
granted by implication, exhaustion, estoppel or otherwise.

**6. Representations and Warranties.** You represent that You are legally entitled to grant the above
licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
have disclosed under Section 3 ). You represent that You have secured permission from Your employer to
make the Submission in cases where Your Submission is made in the course of Your work for Your
employer or Your employer has intellectual property rights in Your Submission by contract or applicable
law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
EXPRESSLY STATED IN SECTIONS 3, 4, AND 6 , THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.

**7. Notice to AvaloniaUI OÜ.** You agree to notify AvaloniaUI OÜ in writing of any facts or
circumstances of which You later become aware that would make Your representations in this
Agreement inaccurate in any respect.

**8. Information about Submissions.** You agree that contributions to Projects and information about
contributions may be maintained indefinitely and disclosed publicly, including Your name and other
information that You submit with Your Submission.

**9. Governing Law/Jurisdiction.** This Agreement is governed by the laws of the Republic of Estonia, and
the parties consent to exclusive jurisdiction and venue in the courts sitting in Talinn,
Estonia. The parties waive all defenses of lack of personal jurisdiction and forum non-conveniens.

**10. Entire Agreement/Assignment.** This Agreement is the entire agreement between the parties, and
supersedes any and all prior agreements, understandings or communications, written or oral, between
the parties relating to the subject matter hereof. This Agreement may be assigned by AvaloniaUI OÜ.


*AvaloniaUI OÜ dedicates this Contribution License Agreement to the public domain according to the Creative Commons CC0 1.*

</details>

**2** out of **3** committers have signed the CLA.
- [x] @jmacato
- [x] @maxkatz6
- [ ] @affederaffe

**Comment by kekekeks** [comment]
The plan is to figure something about CSD support in core avalonia and continue. 

Alternatively we could release it without CSD for usage with embedded systems, wayland seems to be ready enough for those.

**Comment by affederaffe** [comment]
There are just too many unresolved issues, until I have the time and motivation to address those this PR continues to be WIP and thus unmergable from my perspective.
If everything goes well I'll probably have some time in September to work on some of these issues.

**Comment by slackerbob** [comment]
> The plan is to figure something about CSD support in core avalonia and continue.
> 
> Alternatively we could release it without CSD for usage with embedded systems, wayland seems to be ready enough for those.

My understanding is that only GNOME requires CSD, and most other DEs are happy with SSD.

If that's true, then it might make more sense to just default back to X11 for GNOME, and use Wayland on everything else?

**Comment by affederaffe** [comment]
Even though I disagree with GNOME's philosophy to not support SSDs, they are still a major platform we cannot ignore. At least for me, it's full wayland support that works on every platform or none at all.

**Comment by fabiorush92** [comment]
My 2 cents on this topic: as an embedded developer I'm testing Avalonia and I noticed a trend: the latest implementations of Yocto/EmbeddedLinux (one of all for example NXP i.MX8) lack X11 implementation on graphics drivers by specific choice of the OEM. Therefore, at present, DRM is the only way to achieve hardware acceleration, with all the limitations of it (cursor and keyboard for example). In my opinion in the short/medium term Avalonia must have compatibility on Wayland to be 100% competitive on Linux, especially on the industrial/embedded side.

**Comment by Ammar-Sadaoui** [comment]
> Even though I disagree with GNOME's philosophy to not support SSDs, they are still a major platform we cannot ignore. At least for me, it's full wayland support that works on every platform or none at all.

Once again, we should thanks the GNOME developers for sabotaging the Linux desktop.

**Comment by Ammar-Sadaoui** [comment]
> These steps won't be necessary once NWayland is released officially on NuGet.

@affederaffe  is there any alternative for NWayland in NuGet that work with your PR ?

**Comment by RiedleroD** [comment]
> it's full wayland support that works on every platform or none at all

wayland support on KDE, wlroots and hyprland would improve support for those environments while not degrading GNOME's. I for one would highly apprechiate being able to run Avalonia-based programs with proper scaling. Additionally, I predict that GNOME will allow server-side decorations if enough programs exclude them from wayland support because of that.

**Comment by jgcodes2020** [comment]
> > it's full wayland support that works on every platform or none at all
> 
> wayland support on KDE, wlroots and hyprland would improve support for those environments while not degrading GNOME's. I for one would highly apprechiate being able to run Avalonia-based programs with proper scaling. Additionally, I predict that GNOME will allow server-side decorations if enough programs exclude them from wayland support because of that.

GNOME is pretty stubborn with their design philosophy. If anything, you could still rig crude decorations through libdecor until proper decorations are added.

**Comment by Ammar-Sadaoui** [comment]
> > it's full wayland support that works on every platform or none at all
> 
> wayland support on KDE, wlroots and hyprland would improve support for those environments while not degrading GNOME's. I for one would highly apprechiate being able to run Avalonia-based programs with proper scaling. Additionally, I predict that GNOME will allow server-side decorations if enough programs exclude them from wayland support because of that.

nope this will not work with gnome developers

prove : https://gitlab.gnome.org/GNOME/mutter/-/issues/217

**Comment by workgroupengineering** [comment]
If we wait for GNOME to implement SSD to release this PR we will all become centenarians. A fallback method could be added in the AppBuilder that is invoked when the server does not implement SSD. Each developer will implement their own fallback based on their needs

**Comment by Ammar-Sadaoui** [comment]
> If we wait for GNOME to implement SSD to release this PR we will all become centenarians. A fallback method could be added in the AppBuilder that is invoked when the server does not implement SSD. Each developer will implement their own fallback based on their needs

the only way to wayland to improve is by everyone adopting it and start report issues and complain about missing features and demanding alternative

**Comment by Ammar-Sadaoui** [comment]
Do you have any plans to support [libdecor ](https://gitlab.freedesktop.org/libdecor/libdecor) for handling SSD-less compositors, like Mutter from GNOME?

Using XWayland in a Wayland environment has some drawbacks :

Performance Overhead: Adds extra processing, leading to potential performance drops.
Inconsistent Behavior: X11 apps might not integrate seamlessly with Wayland features.
Limited Features: Doesn’t fully support Wayland’s advanced features like high-DPI scaling.
Increased Complexity: Adds another layer, complicating system debugging.
Security Concerns: Inherits some security issues from X11.
Resource Usage: Consumes extra memory and CPU.
Incompatibility: Some Wayland features may not work well with XWayland apps.

**Comment by DocBrown101** [comment]
> Do you have any plans to support [libdecor ](https://gitlab.freedesktop.org/libdecor/libdecor) for handling SSD-less compositors, like Mutter from GNOME?
> 
> Using XWayland in a Wayland environment has some drawbacks :
> 
> Performance Overhead: Adds extra processing, leading to potential performance drops. Inconsistent Behavior: X11 apps might not integrate seamlessly with Wayland features. Limited Features: Doesn’t fully support Wayland’s advanced features like high-DPI scaling. Increased Complexity: Adds another layer, complicating system debugging. Security Concerns: Inherits some security issues from X11. Resource Usage: Consumes extra memory and CPU. Incompatibility: Some Wayland features may not work well with XWayland apps.

And drag and drop does not work with with Wayland apps!

**Comment by Ammar-Sadaoui** [comment]
> And drag and drop does not work with with Wayland apps!

When running X11 applications on Wayland via XWayland, drag-and-drop may not work as expected due to the compatibility layer's limitations. This is because XWayland translates between X11 and Wayland, and some operations (like drag-and-drop) might not translate perfectly.

and even in native wayland app its sill not wayland fault, 
 The Wayland compositor (like GNOME's Mutter, KDE's KWin, etc.) is responsible for managing drag-and-drop actions. If the compositor does not fully support or properly implement drag-and-drop functionality, it can cause issues.
 
 sometime its app fault : Many applications were originally developed for X11, and not all of them have been fully adapted to work with Wayland. This can lead to missing or broken functionality, including drag-and-drop
 
plus : The [Wayland protocol](https://wayland.freedesktop.org/docs/html/ch04.html#:~:text=The%20Wayland%20protocol%20provides%20clients,see%20it%20as%20wl_data_offer%20object.) provides clients a mechanism for sharing data that allows the implementation of copy-paste and drag-and-drop. 

maybe its your fault : dont use or support desktop that dont have implementation for standards like gnome or anything based on it and change to modern desktop for better and acceptable experience  

ps: cosmic desktop is still in Alpha but they have good devs try it and make issuse if you find bug or problem be change you want to see in ~~the world~~  the linux desktop

**Comment by affederaffe** [comment]
> Do you have any plans to support [libdecor ](https://gitlab.freedesktop.org/libdecor/libdecor) for handling SSD-less compositors, like Mutter from GNOME?
> 
> Using XWayland in a Wayland environment has some drawbacks :
> 
> Performance Overhead: Adds extra processing, leading to potential performance drops.
> Inconsistent Behavior: X11 apps might not integrate seamlessly with Wayland features.
> Limited Features: Doesn’t fully support Wayland’s advanced features like high-DPI scaling.
> Increased Complexity: Adds another layer, complicating system debugging.
> Security Concerns: Inherits some security issues from X11.
> Resource Usage: Consumes extra memory and CPU.
> Incompatibility: Some Wayland features may not work well with XWayland apps.

Using libdecor would pull in a Cairo/Gtk dependency, which is something I'm opposed to. I looked into using the current icon theme for drawing caption buttons though and at the moment this seems doable.

**Comment by amayra** [comment]
> Using libdecor would pull in a Cairo/Gtk dependency, which is something I'm opposed to. I looked into using the current icon theme for drawing caption buttons though and at the moment this seems doable.

are you planning to support Client-side decoration (CSD) just for wayland ?

**Comment by IsaacMarovitz** [comment]
> Using libdecor would pull in a Cairo/Gtk dependency, which is something I'm opposed to. I looked into using the current icon theme for drawing caption buttons though and at the moment this seems doable.

Interesting (and frustrating) that libdecor goes for Cairo rather than Skia, but yeah I don't see why accomplishing CSD on Mutter would be too difficult. The question is of course, do we bother trying to conform to Mutter's design language or not.

**Comment by IsaacMarovitz** [comment]
> are you planning to support Client-side decoration (CSD) just for wayland ?

Really it would be just adding CSD support for Wayland on Mutter specifically.
Screenshot 2024-09-11 at 14 51 14

X11 supports SSD, and the only major Wayland compositor to not support SSD is Mutter.

**Comment by maxkatz6** [comment]
I am going to close this PR as it's not active and more work is pending from our side (i.e. CSD).
If you want to add any comments on the feature in general, please add them to this issue instead https://github.com/AvaloniaUI/Avalonia/issues/1243

**Comment by kekekeks** [comment]
@affederaffe any objections for incorporating your changes into the original NWayland repo? The latest Plasma can _almost_ properly run on my new laptop (save for some graphical glitches and scrolling being unusable for GTK apps), so I'm planning to tackle some Wayland-related work this summer.

**Comment by affederaffe** [comment]
> @affederaffe any objections for incorporating your changes into the original NWayland repo? The latest Plasma can _almost_ properly run on my new laptop (save for some graphical glitches and scrolling being unusable for GTK apps), so I'm planning to tackle some Wayland-related work this summer.

Absolutely not, please go ahead!
I'd love Avalonia to finally have a native wayland backend now that I don't have that much time myself anymore.

**Pull request closed without merging**